numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,451 @@
+    if (add_loop(umath, "less", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LE, npy_ucs4>;
+    if (add_loop(umath, "less_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GT, npy_ucs4>;
+    if (add_loop(umath, "greater", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GE, npy_ucs4>;
+    if (add_loop(umath, "greater_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+
+    res = 0;
+  finish:
+    Py_DECREF(String);
+    Py_DECREF(Unicode);
+    Py_DECREF(Bool);
+    return res;
+}
+
+
+template <bool rstrip, typename character>
+static PyArrayMethod_StridedLoop *
+get_strided_loop(int comp)
+{
+    if (comp == Py_EQ) {
+        return string_comparison_loop<rstrip, Py_EQ, character>;
+    }
+    else if (comp == Py_NE) {
+        return string_comparison_loop<rstrip, Py_NE, character>;
+    }
+    else if (comp == Py_LT) {
+        return string_comparison_loop<rstrip, Py_LT, character>;
+    }
+    else if (comp == Py_LE) {
+        return string_comparison_loop<rstrip, Py_LE, character>;
+    }
+    else if (comp == Py_GT) {
+        return string_comparison_loop<rstrip, Py_GT, character>;
+    }
+    else if (comp == Py_GE) {
+        return string_comparison_loop<rstrip, Py_GE, character>;
+    }
+    assert(0);
+    return nullptr;
+}
+
+
+/*
+ * This function is used for `compare_chararrays` (and void comparisons
+ * currently).  The first could probably be deprecated.
+ *
+ * The `rstrip` mechanism is presumably for some fortran compat, but the
+ * question is whether it would not be better to have/use `rstrip` on such
+ * an array first...
+ *
+ * NOTE: This function is also used for unstructured voids, this works because
+ *       `npy_byte` works for it.
+ */
+extern "C" {
+    NPY_NO_EXPORT PyObject *
+    _umath_strings_richcompare(
+            PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip);
+}
+
+NPY_NO_EXPORT PyObject *
+_umath_strings_richcompare(
+        PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip)
+{
+    NpyIter *iter = nullptr;
+    PyObject *result = nullptr;
+
+    char **dataptr = nullptr;
+    npy_intp *strides = nullptr;
+    npy_intp *countptr = nullptr;
+    npy_intp size = 0;
+
+    PyArrayMethod_Context context = {
+        .caller = nullptr,
+        .method = nullptr,
+        .descriptors = nullptr,
+    };
+    NpyIter_IterNextFunc *iternext = nullptr;
+
+    npy_uint32 it_flags = (
+            NPY_ITER_EXTERNAL_LOOP | NPY_ITER_ZEROSIZE_OK |
+            NPY_ITER_BUFFERED | NPY_ITER_GROWINNER);
+    npy_uint32 op_flags[3] = {
+            NPY_ITER_READONLY | NPY_ITER_ALIGNED,
+            NPY_ITER_READONLY | NPY_ITER_ALIGNED,
+            NPY_ITER_WRITEONLY | NPY_ITER_ALLOCATE | NPY_ITER_ALIGNED};
+
+    PyArrayMethod_StridedLoop *strided_loop = nullptr;
+    NPY_BEGIN_THREADS_DEF;
```

Comment:
C++ forced me to move all the inits up, or it didn't want the goto, another C99 feature not included in C11?

Reply:
I wonder if flagging these variables as `constexpr` changes something wrt. `goto` ?

Reply:
Just tried for the flags, and it didn't seem to work.